### PR TITLE
Fix mypy errors attributed to `pytorch_lightning.core.datamodule`

### PR DESCRIPTION
**Pull request by JustinGoheen**
## What does this PR do?

Fix mypy errors attributed to `pytorch_lightning.core.datamodule` for issue #13445 

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: src/pytorch_lightning/core/datamodule.py
Comment (JustinGoheen) [inline src/pytorch_lightning/core/datamodule.py:15]: add for ADD_ARGPARSE_RETURN at ln 26

File: src/pytorch_lightning/core/datamodule.py
Comment (JustinGoheen) [inline src/pytorch_lightning/core/datamodule.py:58]: used `Any` instead of `ellipsis` to appease failed pre-commit hook

File: src/pytorch_lightning/core/datamodule.py
Comment (JustinGoheen) [inline src/pytorch_lightning/core/datamodule.py:80]: this isn't correct; I am trying to see if I can force CI code-check to error and provide a hint for the correct return type

File: src/pytorch_lightning/utilities/argparse.py
Comment (JustinGoheen) [inline src/pytorch_lightning/utilities/argparse.py:168]: updated to account for use of `add_argparse_args` in LightningDataModule.

File: src/pytorch_lightning/utilities/argparse.py
Comment (JustinGoheen) [inline src/pytorch_lightning/utilities/argparse.py:222]: addition of LightningDataModule cased union-attr error; ignored for type check case when type is LightningDataModule

**Comment by JustinGoheen** [comment]
potentially blocked by `core.saving` of #13445

File: src/pytorch_lightning/core/saving.py
Comment (JustinGoheen) [inline src/pytorch_lightning/core/saving.py:174]: updated to resolve

```sh
src/pytorch_lightning/core/datamodule.py:248: error: Argument 1 to "_load_from_checkpoint" has incompatible type "Type[LightningDataModule]"; expected "Union[LightningModule, LightningDataModule]"  [arg-type]
src/pytorch_lightning/core/datamodule.py:249: error: Argument 2 to "_load_from_checkpoint" has incompatible type "Union[Union[str, Path], IO[Any]]"; expected "Union[str, IO[Any]]"  [arg-type]
src/pytorch_lightning/core/datamodule.py:251: error: Argument "hparams_file" to "_load_from_checkpoint" has incompatible type "Union[str, Path, None]"; expected "Optional[str]"  [arg-type]
```

**Comment by JustinGoheen** [comment]
I had originally marked this blocked by another contributor's anticipated PR; however, that contributor has not opened a draft after 2 weeks https://github.com/Lightning-AI/lightning/issues/13445#issuecomment-1185436344

File: src/pytorch_lightning/utilities/argparse.py
Comment (JustinGoheen) [inline src/pytorch_lightning/utilities/argparse.py:43]: original error this change is meant to resolve:

```sh
src/pytorch_lightning/core/datamodule.py:98: error: Argument 1 to "from_argparse_args" has incompatible type "Type[LightningDataModule]"; expected "Type[ParseArgparserDataType]"  [arg-type]
```

File: src/pytorch_lightning/utilities/argparse.py
Comment (JustinGoheen) [inline src/pytorch_lightning/utilities/argparse.py:65]: I opted to ignore here rather than attempt to use something like

```python
if isinstance(args, ArgumentParser):
       assert isintance(class, ParseArgparserDataType)
        args = cls.parse_argparser(args)
```

because we know for certain that LightningDataModule is passed in as cls; meaning an assertion that cls must be `ParseArgparserDataType` might prevent the code from working as intended.

File: src/pytorch_lightning/core/datamodule.py
Comment (JustinGoheen) [inline src/pytorch_lightning/core/datamodule.py:162]: flagged for guidance regarding ignore

File: src/pytorch_lightning/core/datamodule.py
Comment (JustinGoheen) [inline src/pytorch_lightning/core/datamodule.py:164]: flagged for guidance regarding ignore

File: src/pytorch_lightning/core/datamodule.py
Comment (JustinGoheen) [inline src/pytorch_lightning/core/datamodule.py:166]: flagged for guidance regarding ignore

File: src/pytorch_lightning/core/datamodule.py
Comment (JustinGoheen) [inline src/pytorch_lightning/core/datamodule.py:168]: flagged for guidance regarding ignore

File: src/pytorch_lightning/core/datamodule.py
Comment (otaj) [inline src/pytorch_lightning/core/datamodule.py:193]: The return type is not correct

File: src/pytorch_lightning/core/datamodule.py
Comment (otaj) [inline src/pytorch_lightning/core/datamodule.py:134]: I don't think `Any` is appropriate here

File: src/pytorch_lightning/core/saving.py
Comment (otaj) [reply src/pytorch_lightning/core/saving.py:174]: This is absolutely not correct - either both should be `Type[...]` or instances

File: src/pytorch_lightning/utilities/argparse.py
Comment (otaj) [reply src/pytorch_lightning/utilities/argparse.py:43]: I have to say, I'm not sure about this

File: src/pytorch_lightning/utilities/argparse.py
Comment (otaj) [inline src/pytorch_lightning/utilities/argparse.py:65]: I have to say I'm not sure about these `# type: ignore`s. I'm gonna have to think about it a bit

File: src/pytorch_lightning/utilities/argparse.py
Comment (otaj) [reply src/pytorch_lightning/utilities/argparse.py:224]: Same comment about not sure about these ignores applies.

**Review by otaj — CHANGES_REQUESTED** [review]

File: pyproject.toml
Comment (JustinGoheen) [reply pyproject.toml:54]: I didn't add this back in; possible it was missed during recent merge conflict resolution.

File: src/pytorch_lightning/core/datamodule.py
Comment (otaj) [inline src/pytorch_lightning/core/datamodule.py:63]:
```suggestion
    name: str
```
We still want to have it typed, but the assignment is not necessary

File: src/pytorch_lightning/core/datamodule.py
Comment (JustinGoheen) [reply src/pytorch_lightning/core/datamodule.py:162]: please see original comment https://github.com/Lightning-AI/lightning/pull/13693#discussion_r933096521

File: src/pytorch_lightning/core/datamodule.py
Comment (JustinGoheen) [reply src/pytorch_lightning/core/datamodule.py:166]: please see original comment https://github.com/Lightning-AI/lightning/pull/13693#discussion_r933096521

File: src/pytorch_lightning/core/datamodule.py
Comment (JustinGoheen) [reply src/pytorch_lightning/core/datamodule.py:164]: please see original comment https://github.com/Lightning-AI/lightning/pull/13693#discussion_r933096521

File: src/pytorch_lightning/core/datamodule.py
Comment (JustinGoheen) [reply src/pytorch_lightning/core/datamodule.py:168]: please see original comment https://github.com/Lightning-AI/lightning/pull/13693#discussion_r933096521

File: src/pytorch_lightning/core/datamodule.py
Comment (JustinGoheen) [reply src/pytorch_lightning/core/datamodule.py:193]: I traced this to `pytorch_lightning.core.saving._load_from_checkpoint`; which calls `pytorch_lightning.core.saving._load_state` ... load_state also appears to have an incorrect return type of Any. Based on [obj be set to cls](https://github.com/Lightning-AI/lightning/blob/cd92e3541eab86a837b07515814c93e0ed1cac9f/src/pytorch_lightning/core/saving.py#L250) in core.saving._load_state, and cls being either `"pl.LightningModule", "pl.LightningDataModule"`, I think that `core.datamodule.load_from_checkpoint` `core.saving._load_from_checkpoint` and `core.saving._load_state` should all have returns of `"pl.LightningModule", "pl.LightningDataModule"` ...

Is that correct?

Comment (otaj) [reply src/pytorch_lightning/core/datamodule.py:193]: That seems correct. Would you like to take `core.saving` to work on that first?

File: src/pytorch_lightning/core/datamodule.py
Comment (otaj) [reply src/pytorch_lightning/core/datamodule.py:162]: I thought mypy is smarter than that :facepalm: Disregard this then

File: src/pytorch_lightning/core/datamodule.py
Comment (JustinGoheen) [reply src/pytorch_lightning/core/datamodule.py:193]: sounds good to me; I'm going to create my working branch for `core.saving` from this branch, and once again label this branch as blocked by saving, if that is okay with you.

File: src/pytorch_lightning/core/datamodule.py
Comment (JustinGoheen) [reply src/pytorch_lightning/core/datamodule.py]: updated in https://github.com/Lightning-AI/lightning/pull/13693/commits/2f6e162ab7e399e38b61d4ff57e2982fe29eceb7

File: src/pytorch_lightning/core/datamodule.py
Comment (JustinGoheen) [reply src/pytorch_lightning/core/datamodule.py]: updated in https://github.com/Lightning-AI/lightning/pull/13693/commits/2f6e162ab7e399e38b61d4ff57e2982fe29eceb7

File: src/pytorch_lightning/core/datamodule.py
Comment (JustinGoheen) [reply src/pytorch_lightning/core/datamodule.py]: updated in https://github.com/Lightning-AI/lightning/pull/13693/commits/2f6e162ab7e399e38b61d4ff57e2982fe29eceb7

File: src/pytorch_lightning/core/datamodule.py
Comment (JustinGoheen) [reply src/pytorch_lightning/core/datamodule.py]: updated in https://github.com/Lightning-AI/lightning/pull/13693/commits/2f6e162ab7e399e38b61d4ff57e2982fe29eceb7

File: src/pytorch_lightning/core/datamodule.py
Comment (JustinGoheen) [reply src/pytorch_lightning/core/datamodule.py]: updated in https://github.com/Lightning-AI/lightning/pull/13693/commits/2f6e162ab7e399e38b61d4ff57e2982fe29eceb7

File: src/pytorch_lightning/core/datamodule.py
Comment (JustinGoheen) [reply src/pytorch_lightning/core/datamodule.py]: updated in https://github.com/Lightning-AI/lightning/pull/13693/commits/2f6e162ab7e399e38b61d4ff57e2982fe29eceb7

File: src/pytorch_lightning/utilities/argparse.py
Comment (JustinGoheen) [reply src/pytorch_lightning/utilities/argparse.py]: @otaj resolved in https://github.com/Lightning-AI/lightning/pull/13693/commits/517576db4d0b1cbda58bc1ac63416436e8ddd198

File: src/pytorch_lightning/utilities/argparse.py
Comment (JustinGoheen) [reply src/pytorch_lightning/utilities/argparse.py]: resolved in https://github.com/Lightning-AI/lightning/pull/13693/commits/517576db4d0b1cbda58bc1ac63416436e8ddd198

**Comment by JustinGoheen** [comment]
@otaj @carmocca @Borda @rohitgr7 

The new, suggested class methods in LightningDataModule are to reconcile errors raised by mypy; however, these errors originate from pl.utilities.argparse. 

While troubleshooting the mypy error:

```sh
src/pytorch_lightning/utilities/argparse.py:65: error: Item "LightningDataModule" of "Union[ParseArgparserDataType, LightningDataModule]" has no attribute "parse_argparser"  [union-attr]
```

It was found that in pl.utilities.argparse, there appears to be an unused base class (ParseArgparserDataType) that was perhaps intended to act as a mixin/utility interface for LightningDataModule and Trainer. This base class originated in https://github.com/Lightning-AI/lightning/pull/8124, as pointed out by Ota in a Slack conversation. 

Should I integrate ParseArgparserDataType in this PR – refactoring both LightningDataModule and Trainer appropriately to inherit from the base class; or should I open a new PR to handle this task?

If this PR should not integrate ParseArgparserDataType, then it is ready for review; otherwise, it should be converted to a draft.

File: src/pytorch_lightning/core/datamodule.py
Comment (carmocca) [inline src/pytorch_lightning/core/datamodule.py:118]: Why is this added?

File: src/pytorch_lightning/core/datamodule.py
Comment (carmocca) [inline src/pytorch_lightning/core/datamodule.py:32]: Isn't this the same? Also I would mark this protected

```suggestion
_ADD_ARGPARSE_RETURN = Union[_ArgumentGroup, ArgumentParser]
```

File: src/pytorch_lightning/utilities/argparse.py
Comment (carmocca) [inline src/pytorch_lightning/utilities/argparse.py:33]: Could this class be removed now?

**Review by carmocca — COMMENTED** [review]
> If this PR should not integrate ParseArgparserDataType

I lean towards this option if we can get rid of it.

File: src/pytorch_lightning/core/datamodule.py
Comment (JustinGoheen) [reply src/pytorch_lightning/core/datamodule.py:118]: mypy raises a union-attr error, as it is necessary in [`utilities.argparse.add_argparse_args`](https://github.com/JustinGoheen/lightning/blob/d2c086b04ad14ec5d6cf63e9487381a96eee1733/src/pytorch_lightning/utilities/argparse.py#L167); however, `get_deprecated_arg_names` isn't a function in `utilities.argparse`, it is a [class method of Trainer](https://github.com/JustinGoheen/lightning/blob/d2c086b04ad14ec5d6cf63e9487381a96eee1733/src/pytorch_lightning/trainer/trainer.py#L2433). 

I think the more correct thing to do, is to move the implementation of `get_deprecated_arg_names` to `utilities.argparse`, and then update `core.datamodule` and `trainer.trainer` accordingly ... do you agree?

File: src/pytorch_lightning/utilities/argparse.py
Comment (JustinGoheen) [reply src/pytorch_lightning/utilities/argparse.py]: removed in https://github.com/Lightning-AI/lightning/pull/13693/commits/850263e941b63a74355a606664a685aee1c66dcd

File: src/pytorch_lightning/core/saving.py
Comment (JustinGoheen) [inline src/pytorch_lightning/core/saving.py:204]: removed my comments from a previous PR. these comments no longer apply

**Comment by JustinGoheen** [comment]
@rohitgr7 what was the intended effect of setting strict to None? strict needs to be a bool, so I am wondering if this needs to be False or left as none and ignore the mypy error https://github.com/Lightning-AI/lightning/blob/d29a552b3c701ebc14d608347c1dbf55c3dfaa6a/src/pytorch_lightning/core/datamodule.py#L245

`_load_from_checkpoint` was updated in the PR for core.saving yesterday.

@awaelchli this will likely affect your update to core.saving as well

File: src/pytorch_lightning/core/datamodule.py
Comment (carmocca) [reply src/pytorch_lightning/core/datamodule.py:117]: I agree. But let's do that in a PR before this one or a PR after.

Comment (JustinGoheen) [reply src/pytorch_lightning/core/datamodule.py:117]: Sounds good to me. I'll open a new issue referring to this thread, and will follow up with a new PR after this one is merged.

**Comment by rohitgr7** [comment]
@JustinGoheen added a commit to resolve the issue

**Comment by JustinGoheen** [comment]
we probably shouldn't change core.saving since Adrian is working a separate PR to update annotations for .saving

File: src/pytorch_lightning/utilities/argparse.py
Comment (JustinGoheen) [inline src/pytorch_lightning/utilities/argparse.py:124]: note to self: remove this Any and check mypy locallly

Comment (JustinGoheen) [reply src/pytorch_lightning/utilities/argparse.py]: this Any was added to avoid the following error

```sh
src/pytorch_lightning/utilities/argparse.py:222: error: Argument 1 to "get_init_arguments_and_types" has incompatible type "Union[Type[LightningDataModule], Type[Trainer], Callable[[LightningDataModule], None], Callable[[Trainer, Union[Logger, Iterable[Logger], bool], bool, Union[List[Callback], Callback, None], Optional[str], Union[int, float, None], Optional[str], int, Optional[int], Union[List[int], str, int, None], Union[List[int], str, int, None], bool, Union[List[int], str, int, None], Optional[int], bool, Union[int, float], Union[int, float, str], Optional[int], Union[int, bool], Union[int, Dict[int, int], None], Optional[int], Optional[int], int, Optional[int], Union[str, timedelta, Dict[str, int], None], Union[int, float, None], Union[int, float, None], Union[int, float, None], Union[int, float, None], Union[int, float, None], int, Union[str, Accelerator, None], Union[str, Strategy, None], bool, Union[int, str], bool, Optional[str], int, Union[Path, str, None], Union[Profiler, str, None], Optional[bool], Union[bool, Literal['warn'], None], int, Union[bool, str], bool, bool, Union[str, bool], Union[Strategy, PrecisionPlugin, ClusterEnvironment, CheckpointIO, LayerSync, str, List[Union[Union[Strategy, PrecisionPlugin, ClusterEnvironment, CheckpointIO, LayerSync], str]], None], str, Optional[str], bool, str], None]]"; expected "Union[Type[LightningDataModule], Type[Trainer]]"  [arg-type]
```

I ignored that error in favor of removing the Any, in commit https://github.com/Lightning-AI/lightning/pull/13693/commits/cc55ab96c3ac5e7399c29bd77ab8b6ac8217eb2d

**Comment by JustinGoheen** [comment]
apologies, I'm no longer able to keep this PR up to date.

**Review by rohitgr7 — APPROVED** [review]

File: src/pytorch_lightning/core/saving.py
Comment (carmocca) [inline src/pytorch_lightning/core/saving.py:211]: Why did this become optional?

**Review by carmocca — APPROVED** [review]

File: src/pytorch_lightning/core/saving.py
Comment (rohitgr7) [reply src/pytorch_lightning/core/saving.py:211]: LDM doesn't need `strict` argument and we use the same method for LM and LDM. It's a protected method so not affecting user.

**Review by otaj — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Pull request merged**
